src/query_plan/operation.rs

```diff
@@ -1735,6 +2370,248 @@ pub(crate) fn merge_selection_sets(
+    fn default() -> NamedFragments {
+        NamedFragments {
+            fragments: Arc::new(IndexMap::new()),
+        }
+    }
+
+    pub(crate) fn new(
+        fragments: &IndexMap<Name, Node<Fragment>>,
+        schema: &ValidFederationSchema,
+    ) -> NamedFragments {
+        // JS PORT - In order to normalize Fragments we need to process them in dependency order.
+        //
+        // In JS implementation map_in_dependency_order method was called when rebasing/filtering/expanding selection sets.
+        // Since resulting `IndexMap` of `NormalizedFragments` will be already sorted, we only need to map it once
+        // when creating the `NamedFragments`.
+        NamedFragments::map_in_dependency_order(fragments, schema)
+    }
+
+    pub(crate) fn is_empty(&self) -> bool {
+        self.fragments.len() == 0
+    }
+
+    pub(crate) fn size(&self) -> usize {
+        self.fragments.len()
+    }
+
+    pub(crate) fn insert(&mut self, fragment: NormalizedFragment) {
+        Arc::make_mut(&mut self.fragments).insert(fragment.name.clone(), Node::new(fragment));
+    }
+
+    pub(crate) fn try_insert(
+        &mut self,
+        fragment: NormalizedFragment,
+    ) -> Result<(), FederationError> {
+        match Arc::make_mut(&mut self.fragments).entry(fragment.name.clone()) {
+            indexmap::map::Entry::Occupied(_) => {
+                Err(FederationError::internal("Duplicate fragment name"))
+            }
+            indexmap::map::Entry::Vacant(entry) => {
+                let _ = entry.insert(Node::new(fragment));
+                Ok(())
+            }
+        }
+    }
+
+    pub(crate) fn get(&self, name: &Name) -> Option<Node<NormalizedFragment>> {
+        self.fragments.get(name).cloned()
+    }
+
+    pub(crate) fn contains(&self, name: &Name) -> bool {
+        self.fragments.contains_key(name)
+    }
+
+    /**
+     * Collect the usages of fragments that are used within the selection of other fragments.
+     */
+    pub(crate) fn collect_used_fragment_names(&self, aggregator: &mut HashMap<Name, i32>) {
+        self.fragments.iter().for_each(|(_, fragment)| {
+            fragment
+                .selection_set
+                .collect_used_fragment_names(aggregator)
+        });
+    }
+
+    /// JS PORT NOTE: In JS implementation this method was called when rebasing/filtering/expanding selection sets.
+    /// JS PORT NOTE: In JS implementation this method was potentially returning `undefined`, in order to simplify the code
+    /// we will always return `NamedFragments` even if they are empty.
+    ///
+    /// We normalize passed in fragments in their dependency order, i.e. if a fragment A uses another fragment B, then we will
+    /// normalize B _before_ attempting to normalize A. Normalized fragments have access to previously normalized fragments.
+    fn map_in_dependency_order(
+        fragments: &IndexMap<Name, Node<Fragment>>,
+        schema: &ValidFederationSchema,
+    ) -> NamedFragments {
```

Comment:
Yes it is on purpose -> as we need to normalize fragments in the dependency order, this method is now invoked on `NamedFragments` creation (see [link](https://github.com/apollographql/federation-next/blob/39a77d66fbedcb2190d8000a8578e2ca3662fdf6/src/query_plan/operation.rs#L2422-L2432)). Since our `IndexMap` will be sorted there is no need to keep rerunning it.

I'll rename the method to `normalize_in_dependency_order` (or something similar)